src/workerd/api/node/filesystem.h

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+// Copyright Joyent and Node contributors. All rights reserved. MIT license.
+#pragma once
+
+#include <workerd/jsg/buffersource.h>
+#include <workerd/jsg/jsg.h>
+#include <workerd/jsg/jsvalue.h>
+#include <workerd/jsg/url.h>
+
+#include <kj/string.h>
+
+namespace workerd::api::node {
+
+class FilesystemUtil final: public jsg::Object {
+ public:
+  FilesystemUtil() = default;
+  FilesystemUtil(jsg::Lock&, const jsg::Url&) {}
+
+  enum class NodeType : kj::uint { FILE, DIRECTORY };
+
+  class Node: public jsg::Object {
+    Node(kj::String name, kj::Path path, kj::Date modifiedAt, kj::Date createdAt, NodeType type)
+        : name(kj::mv(name)),
+          path(kj::mv(path)),
+          modifiedAt(kj::mv(modifiedAt)),
+          createdAt(kj::mv(createdAt)),
+          type(type) {};
+
+    virtual bool isReadable() const = 0;
+    virtual bool isWritable() const = 0;
+    virtual bool isAsyncOnly() const = 0;
+    virtual bool isSyncOnly() const = 0;
+
+    kj::StringPtr getName() const {
+      return name;
+    }
+    kj::StringPtr getPath() const {
+      return path.toString();
+    }
+    jsg::JsDate getModifiedAt() const {
+      return jsg::JsDate(modifiedAt);
+    }
+    jsg::JsDate getCreatedAt() const {
+      return jsg::JsDate(createdAt);
+    }
+
+    bool isDirectory() const {
+      return type == NodeType::DIRECTORY;
+    }
+    bool isFile() const {
+      return type == NodeType::FILE;
+    }
+
+    JSG_RESOURCE_TYPE(Node) {
+      JSG_READONLY_INSTANCE_PROPERTY(readable, isReadable);
+      JSG_READONLY_INSTANCE_PROPERTY(writable, isWritable);
+      JSG_READONLY_INSTANCE_PROPERTY(asyncOnly, isAsyncOnly);
+      JSG_READONLY_INSTANCE_PROPERTY(syncOnly, isSyncOnly);
```

Comment:
Since these are virtual you can't do it like this. Instead, you'll need anything that inherits from `Node` here to have these mapped in its own `JSG_RESOURCE_TYPE` block... e.g. something like

```cpp
class Foo: public jsg::Object {
 public:
  virtual void doFoo() = 0;
}

class MyFoo: public Foo {
 public:
  void doFoo() override {}

  JSG_RESOURCE_TYPE(MyFoo) {
    JSG_METHOD(doFoo);    
  }
}
```